In [173]:
import requests
import urllib.request
import time
from bs4 import BeautifulSoup
import numpy as np
import pandas as pd
from urllib.request import urlopen
import geocoder


pd.set_option('display.max_columns', 500)
pd.set_option('display.max_rows', 100)

print("Libraries imported")

Libraries imported


## Part 1: Fetch the Postal Codes, Boroughs and Neighborhoods of Toronto

#### 1.1 Read the HTML page and initialize a BeautifulSoup object with its contents

In [174]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
html = urlopen(url) 
soup = BeautifulSoup(html, 'html.parser')

#### 1.2 Find all tables in the page

In [175]:
tables = soup.find_all('table')

#### 1.3 Iterate through the cells of the first table and retrieve the postal codes, boroughs and neighborhoods

In [176]:
#Create array to hold the data we extract
postal_codes = []
boroughs = []
neighborhoods = []

for table in tables:
    rows = table.find_all('tr')
    
    for row in rows:
        cells = row.find_all('td')
        
        if len(cells) > 1:
            postal_code = cells[0]
            postal_codes.append(postal_code.text.replace('\n', '')) # remove any trailing new line character
            
            borough = cells[1]
            boroughs.append(borough.text.replace('\n', ''))
            
            neighborhood = cells[2]
            neighborhoods.append(neighborhood.text.replace('\n', ''))
    
    break # we only need to parse the first table         

#### 1.4 Transform the data to a dataframe

In [177]:
df = pd.DataFrame({'PostalCode': postal_codes,
                   'Borough': boroughs,
                   'Neighborhood': neighborhoods})

df

PostalCode           Borough  \
0          M1A      Not assigned   
1          M2A      Not assigned   
2          M3A        North York   
3          M4A        North York   
4          M5A  Downtown Toronto   
..         ...               ...   
175        M5Z      Not assigned   
176        M6Z      Not assigned   
177        M7Z      Not assigned   
178        M8Z         Etobicoke   
179        M9Z      Not assigned   

                                          Neighborhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

In [178]:
df = df[df['Borough'] != "Not assigned"] #remove any rows without Borough

df

PostalCode           Borough  \
2          M3A        North York   
3          M4A        North York   
4          M5A  Downtown Toronto   
5          M6A        North York   
6          M7A  Downtown Toronto   
..         ...               ...   
160        M8X         Etobicoke   
165        M4Y  Downtown Toronto   
168        M7Y      East Toronto   
169        M8Y         Etobicoke   
178        M8Z         Etobicoke   

                                          Neighborhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

#### 1.5 Finally keep the postal codes we need and drop the rest

In [179]:
#selected_postal_codes = ['M5G', 'M2H', 'M4B', 'M1J', 'M4G', 'M4M', 'M1R', 'M9V', 'M9L', 'M5V', 'M1B', 'M5A']
#df = df[df['PostalCode'].isin(selected_postal_codes)]
#df_toronto = df.reset_index(drop=True)

df_toronto = df
df_toronto

PostalCode           Borough  \
2          M3A        North York   
3          M4A        North York   
4          M5A  Downtown Toronto   
5          M6A        North York   
6          M7A  Downtown Toronto   
..         ...               ...   
160        M8X         Etobicoke   
165        M4Y  Downtown Toronto   
168        M7Y      East Toronto   
169        M8Y         Etobicoke   
178        M8Z         Etobicoke   

                                          Neighborhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [180]:
df_toronto.shape

(103, 3)

## Part 2: Fetch the latitude and longitude of the neighborhoods


In [181]:
# Code to split each row that has more than one neighborhood for a postal code into separate rows, not needed anymore

# df_toronto = pd.DataFrame(columns = ['PostalCode', 'Borough', 'Neighborhood']) 
# df_toronto.dtypes

# for index, row in df.iterrows():
#     neighborhoods = row['Neighborhood'].split(',')
#     if len(neighborhoods) > 1 :
#         for neighborhood in neighborhoods:
#             new_row = {'PostalCode': row['PostalCode'], 'Borough': row['Borough'], 'Neighborhood': neighborhood}
#             df_toronto = df_toronto.append(new_row, ignore_index=True)
#     else:
#         new_row = {'PostalCode': row['PostalCode'], 'Borough': row['Borough'], 'Neighborhood': row['Neighborhood']}
#         df_toronto = df_toronto.append(new_row, ignore_index=True)

# df_toronto

#### 2.1 Load geospatial coordinates for Toronto from CSV file

In [182]:
df_location = pd.read_csv("Geospatial_Coordinates.csv")
df_location.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

#### 2.2 Loop on each postal code to get its latitude and longitude

In [183]:
latitudes = []
longitudes = []

for index, row in df.iterrows():
    postal_code = row['PostalCode']

    location = df_location.loc[df_location['Postal Code'] == postal_code]

    latitude = location.iloc[0]['Latitude']
    longitude = location.iloc[0]['Longitude']

    latitudes.append(latitude)
    longitudes.append(longitude)

df_toronto['Latitude'] = latitudes
df_toronto['Longitude'] = longitudes

df_toronto

PostalCode           Borough  \
2          M3A        North York   
3          M4A        North York   
4          M5A  Downtown Toronto   
5          M6A        North York   
6          M7A  Downtown Toronto   
..         ...               ...   
160        M8X         Etobicoke   
165        M4Y  Downtown Toronto   
168        M7Y      East Toronto   
169        M8Y         Etobicoke   
178        M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
2                                            Parkwoods  43.753259 -79.329656  
3                                     Victoria Village  43.725882 -79.315572  
4                            Regent Park, Harbourfront  43.654260 -79.360636  
5                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
6          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
160      The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
165                               Church and Wellesley  43.665860 -79.383160  
168  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
178  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

## Part 3: Explore and Cluster Toronto neighborhoods

In [184]:
import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address to latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


## 3.1. Visualize Toronto neighborhoods

In [185]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="to_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [186]:
# create map of New York using latitude and longitude values
map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_toronto['Latitude'], df_toronto['Longitude'], df_toronto['Borough'], df_toronto['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

#### Define Foursquare Credentials and Version

In [195]:
CLIENT_ID = '00000000000000000000000' # your Foursquare ID
CLIENT_SECRET = '000000000000000000000000' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: G3I1S3STTCK0OMQXUGZPMWJZ2LAZN0AUYT0NB23OBVG1TJEU
CLIENT_SECRET:NOOBRTQ22EA3OPFVBE5ULOXSXX3MTMHKDQLWLQMT3M0ACUXR


In [196]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

## 3.2. Explore Toronto neighborhoods

#### Let's create a function to explore all the neighborhoods in Toronto

In [198]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        #print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']

        print('{}: {} venues'.format(name, len(results)))

        
        # return only relevant information for each nearby venue

        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [199]:
neighborhood_names = df_toronto.loc[:, 'Neighborhood'] # neighborhood names
neighborhood_latitudes = df_toronto.loc[:, 'Latitude'] # neighborhood latitude values
neighborhood_longitudes = df_toronto.loc[:, 'Longitude'] # neighborhood longitude values

toronto_venues = getNearbyVenues(neighborhood_names, neighborhood_latitudes, neighborhood_longitudes)

toronto_venues.head()

Parkwoods: 3 venues
Victoria Village: 5 venues
Regent Park, Harbourfront: 48 venues
Lawrence Manor, Lawrence Heights: 14 venues
Queen's Park, Ontario Provincial Government: 32 venues
Islington Avenue, Humber Valley Village: 0 venues
Malvern, Rouge: 2 venues
Don Mills: 5 venues
Parkview Hill, Woodbine Gardens: 13 venues
Garden District, Ryerson: 100 venues
Glencairn: 4 venues
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale: 0 venues
Rouge Hill, Port Union, Highland Creek: 2 venues
Don Mills: 19 venues
Woodbine Heights: 5 venues
St. James Town: 80 venues
Humewood-Cedarvale: 4 venues
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood: 8 venues
Guildwood, Morningside, West Hill: 7 venues
The Beaches: 5 venues
Berczy Park: 58 venues
Caledonia-Fairbanks: 4 venues
Woburn: 4 venues
Leaside: 32 venues
Central Bay Street: 64 venues
Christie: 16 venues
Cedarbrae: 8 venues
Hillcrest Village: 5 venues
Bathurst Manor, Wilson Heights, Downsview North: 22 venues
Thorn

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2         Parkwoods              43.753259              -79.329656   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1         Brookbanks Pool       43.751389       -79.332184   
2           Variety Store       43.751974       -79.333114   
3  Victoria Village Arena       43.723481       -79.315635   
4               Portugril       43.725819       -79.312785   

          Venue Category  
0                   Park  
1                   Pool  
2      Food & Drink Shop  
3           Hockey Arena  
4  Portuguese Restaurant

## 3.3 Analyze neighborhoods Venues

In [200]:
print(toronto_venues.shape)
print(len(toronto_venues['Neighborhood'].unique()))
toronto_venues.head()

(2122, 7)
96


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2         Parkwoods              43.753259              -79.329656   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1         Brookbanks Pool       43.751389       -79.332184   
2           Variety Store       43.751974       -79.333114   
3  Victoria Village Arena       43.723481       -79.315635   
4               Portugril       43.725819       -79.312785   

          Venue Category  
0                   Park  
1                   Pool  
2      Food & Drink Shop  
3           Hockey Arena  
4  Portuguese Restaurant

In [201]:
toronto_venues.shape

(2122, 7)

In [202]:
toronto_venues.groupby('Neighborhood').count() # how many venues were retruned for each postal code

16              16   
Church and Wellesley                                   80              80   
Clarks Corners, Tam O'Shanter, Sullivan                11              11   
Cliffside, Cliffcrest, Scarborough Village West         2               2   
Commerce Court, Victoria Hotel                        100             100   
Davisville                                             33              33   
Davisville North                                        9               9   
Del Ray, Mount Dennis, Keelsdale and Silverthorn        5               5   
Don Mills                                              24              24   
Dorset Park, Wexford Heights, Scarborough Town ...      6               6   
Downsview                                              15              15   
Dufferin, Dovercourt Village                           17              17   
East Toronto, Broadview North (Old East York)           2               2   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...      8               8   
Fairview, Henry Farm, Oriole                           70              70   
First Canadian Place, Underground city                100             100   
Forest Hill North & West, Forest Hill Road Park         4               4   
Garden District, Ryerson                              100             100   
Glencairn                                               4               4   
Golden Mile, Clairlea, Oakridge                         9               9   
Guildwood, Morningside, West Hill                       7               7   
Harbourfront East, Union Station, Toronto Islands     100             100   
High Park, The Junction South                          24              24   
Hillcrest Village                                       5               5   
Humber Summit                                           2               2   
Humberlea, Emery                                        1               1   
Humewood-Cedarvale                                      4               4   
India Bazaar, The Beaches West                         20              20   
Kennedy Park, Ionview, East Birchmount Park             4               4   
Kensington Market, Chinatown, Grange Park              64              64   
Kingsview Village, St. Phillips, Martin Grove G...      4               4   
Lawrence Manor, Lawrence Heights                       14              14   
Lawrence Park                                           4               4   
Leaside                                                32              32   
Little Portugal, Trinity                               43              43   
Malvern, Rouge                                          2               2   
Milliken, Agincourt North, Steeles East, L'Amor...      3               3   
Mimico NW, The Queensway West, South of Bloor, ...     12              12   
Moore Park, Summerhill East                             2               2   
New Toronto, Mimico South, Humber Bay Shores           12              12   
North Park, Maple Leaf Park, Upwood Park                4               4   
North Toronto West,  Lawrence Park                     19              19   
Northwest, West Humber - Clairville                     3               3   
Northwood Park, York University                         6               6   
Old Mill South, King's Mill Park, Sunnylea, Hum...      1               1   
Parkdale, Roncesvalles                                 14              14   
Parkview Hill, Woodbine Gardens                        13              13   
Parkwoods                                               3               3   
Queen's Park, Ontario Provincial Government            32              32   
Regent Park, Harbourfront                              48              48   
Richmond, Adelaide, King                               96              96   
Rosedale                                                4               4   
Roselawn                                           

Note that we only have 96 neighborhoods with venues. Foursquare did not return any venues for some postal codes

#### Let's find out how many unique categories can be curated from all the returned venues

In [203]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 267 uniques categories.


In [204]:
toronto_venues[toronto_venues['Venue Category'] == "Neighborhood"]

Neighborhood  \
354                                         The Beaches   
596                            Richmond, Adelaide, King   
782   Harbourfront East, Union Station, Toronto Islands   
1255                                    Studio District   

      Neighborhood Latitude  Neighborhood Longitude             Venue  \
354               43.676357              -79.293031     Upper Beaches   
596               43.650571              -79.384568  Downtown Toronto   
782               43.640816              -79.381752      Harbourfront   
1255              43.659526              -79.340923       Leslieville   

      Venue Latitude  Venue Longitude Venue Category  
354        43.680563       -79.292869   Neighborhood  
596        43.653232       -79.385296   Neighborhood  
782        43.639526       -79.380688   Neighborhood  
1255       43.662070       -79.337856   Neighborhood

In [205]:
# drop rows containing Venue Category = "Neighborhood", it has no meaning!
toronto_venues = toronto_venues[toronto_venues['Venue Category'] != "Neighborhood"]

#### Now create one hot encoded data frame with the 267 categories for each neighborhood

In [206]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

toronto_onehot.head()

0            0   
4                      0                   0                  0            0   

   College Rec Center  College Stadium  Colombian Restaurant  \
0                   0                0                     0   
1                   0                0                     0   
2                   0                0                     0   
3                   0                0                     0   
4                   0                0                     0   

   Comfort Food Restaurant  Comic Shop  Concert Hall  \
0                        0           0             0   
1                        0           0             0   
2                        0           0             0   
3                        0           0             0   
4                        0           0             0   

   Construction & Landscaping  Convenience Store  Cosmetics Shop  \
0                           0                  0               0   
1                           0                  0               0   
2                           0                  0               0   
3                           0                  0               0   
4                           0                  0               0   

   Coworking Space  Creperie  Cuban Restaurant  Cupcake Shop  Curling Ice  \
0                0         0                 0             0            0   
1                0         0                 0             0            0   
2                0         0                 0             0            0   
3                0         0                 0             0            0   
4                0         0                 0             0            0   

   Dance Studio  Deli / Bodega  Department Store  Dessert Shop  \
0             0              0                 0             0   
1             0              0                 0             0   
2             0              0                 0             0   
3             0              0                 0             0   
4             0              0                 0             0   

   Dim Sum Restaurant  Diner  Discount Store  Distribution Center  Dog Run  \
0                   0      0               0                    0        0   
1                   0      0               0                    0        0   
2                   0      0               0                    0        0   
3                   0      0               0                    0        0   
4                   0      0               0                    0        0   

   Doner Restaurant  Donut Shop  Drugstore  Eastern European Restaurant  \
0                 0           0          0                            0   
1                 0           0          0                            0   
2                 0           0          0                            0   
3                 0           0          0                            0   
4                 0           0          0                            0   

   Electronics Store  Escape Room  Ethiopian Restaurant  Event Space  \
0                  0            0                     0            0   
1                  0            0                     0            0   
2                  0            0                     0            0   
3                  0            0                     0            0   
4                  0            0                     0            0   

   Falafel Restaurant  Farm  Farmers Market  Fast Food Restaurant  Field  \
0                   0     0               0                     0      0   
1                   0     0               0                     0      0   
2                   0     0               0                     0      0   
3                   0     0               0                     0      0   
4                   0     0               0                     0      0   

   Filipino Restaurant  Fish & Chips Shop  Fish Market  Flea Market  \
0                    0       

In [207]:
# Re-add Neighborhood column and make it the first column
columns_names = list(toronto_onehot.columns.values)

toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood']

columns_names.insert(0, 'Neighborhood')

toronto_onehot = toronto_onehot[columns_names]

toronto_onehot.head()

0            0   
4                      0                   0                  0            0   

   College Rec Center  College Stadium  Colombian Restaurant  \
0                   0                0                     0   
1                   0                0                     0   
2                   0                0                     0   
3                   0                0                     0   
4                   0                0                     0   

   Comfort Food Restaurant  Comic Shop  Concert Hall  \
0                        0           0             0   
1                        0           0             0   
2                        0           0             0   
3                        0           0             0   
4                        0           0             0   

   Construction & Landscaping  Convenience Store  Cosmetics Shop  \
0                           0                  0               0   
1                           0                  0               0   
2                           0                  0               0   
3                           0                  0               0   
4                           0                  0               0   

   Coworking Space  Creperie  Cuban Restaurant  Cupcake Shop  Curling Ice  \
0                0         0                 0             0            0   
1                0         0                 0             0            0   
2                0         0                 0             0            0   
3                0         0                 0             0            0   
4                0         0                 0             0            0   

   Dance Studio  Deli / Bodega  Department Store  Dessert Shop  \
0             0              0                 0             0   
1             0              0                 0             0   
2             0              0                 0             0   
3             0              0                 0             0   
4             0              0                 0             0   

   Dim Sum Restaurant  Diner  Discount Store  Distribution Center  Dog Run  \
0                   0      0               0                    0        0   
1                   0      0               0                    0        0   
2                   0      0               0                    0        0   
3                   0      0               0                    0        0   
4                   0      0               0                    0        0   

   Doner Restaurant  Donut Shop  Drugstore  Eastern European Restaurant  \
0                 0           0          0                            0   
1                 0           0          0                            0   
2                 0           0          0                            0   
3                 0           0          0                            0   
4                 0           0          0                            0   

   Electronics Store  Escape Room  Ethiopian Restaurant  Event Space  \
0                  0            0                     0            0   
1                  0            0                     0            0   
2                  0            0                     0            0   
3                  0            0                     0            0   
4                  0            0                     0            0   

   Falafel Restaurant  Farm  Farmers Market  Fast Food Restaurant  Field  \
0                   0     0               0                     0      0   
1                   0     0               0                     0      0   
2                   0     0               0                     0      0   
3                   0     0               0                     0      0   
4                   0     0               0                     0      0   

   Filipino Restaurant  Fish & Chips Shop  Fish Market  Flea Market  \
0                    0       

In [208]:
toronto_onehot.shape

(2118, 267)

In [209]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

0.0000            0.000000          0.000000  0.000000   
94            0.0000            0.000000          0.000000  0.000000   
95            0.0000            0.000000          0.000000  0.000000   

    Train Station  Truck Stop  Turkish Restaurant  \
0        0.000000    0.000000                 0.0   
1        0.000000    0.000000                 0.0   
2        0.000000    0.000000                 0.0   
3        0.000000    0.000000                 0.0   
4        0.000000    0.000000                 0.0   
5        0.000000    0.000000                 0.0   
6        0.000000    0.000000                 0.0   
7        0.000000    0.000000                 0.0   
8        0.000000    0.000000                 0.0   
9        0.000000    0.000000                 0.0   
10       0.000000    0.000000                 0.0   
11       0.000000    0.000000                 0.0   
12       0.000000    0.000000                 0.0   
13       0.000000    0.000000                 0.0   
14       0.000000    0.000000                 0.0   
15       0.000000    0.000000                 0.0   
16       0.000000    0.000000                 0.0   
17       0.000000    0.000000                 0.0   
18       0.000000    0.000000                 0.0   
19       0.000000    0.000000                 0.0   
20       0.000000    0.000000                 0.0   
21       0.000000    0.000000                 0.2   
22       0.000000    0.000000                 0.0   
23       0.000000    0.000000                 0.0   
24       0.000000    0.000000                 0.0   
25       0.000000    0.000000                 0.0   
26       0.000000    0.000000                 0.0   
27       0.000000    0.000000                 0.0   
28       0.000000    0.000000                 0.0   
29       0.010000    0.000000                 0.0   
30       0.000000    0.000000                 0.0   
31       0.000000    0.000000                 0.0   
32       0.000000    0.000000                 0.0   
33       0.000000    0.000000                 0.0   
34       0.000000    0.000000                 0.0   
35       0.010101    0.000000                 0.0   
36       0.000000    0.000000                 0.0   
37       0.000000    0.000000                 0.0   
38       0.000000    0.000000                 0.0   
39       0.000000    0.000000                 0.0   
40       0.000000    0.000000                 0.0   
41       0.000000    0.000000                 0.0   
42       0.000000    0.000000                 0.0   
43       0.000000    0.000000                 0.0   
44       0.000000    0.000000                 0.0   
45       0.000000    0.000000                 0.0   
46       0.000000    0.000000                 0.0   
47       0.000000    0.000000                 0.0   
48       0.000000    0.000000                 0.0   
49       0.000000    0.000000                 0.0   
50       0.000000    0.000000                 0.0   
51       0.000000    0.000000                 0.0   
52       0.000000    0.000000                 0.0   
53       0.000000    0.000000                 0.0   
54       0.000000    0.000000                 0.0   
55       0.000000    0.000000                 0.0   
56       0.000000    0.333333                 0.0   
57       0.000000    0.000000                 0.0   
58       0.000000    0.000000                 0.0   
59       0.000000    0.000000                 0.0   
60       0.000000    0.000000                 0.0   
61       0.000000    0.000000                 0.0   
62       0.000000    0.000000                 0.0   
63       0.000000    0.000000                 0.0   
64       0.000000    0.000000                 0.0   
65       0.000000    0.000000                 0.0   
66       0.000000    0.000000                 0.0   
67       0.000000    0.000000                 0.0   
68       0.000000    0.000000                 0.0   
69       0.000000    0.000000                 0.0   
70       0.000000    0.000000     

In [210]:
toronto_grouped.shape

(96, 267)

In [211]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [212]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood      1st Most Common Venue  \
0                                        Agincourt  Latin American Restaurant   
1                           Alderwood, Long Branch                Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                       Bank   
3                                  Bayview Village                       Café   
4                Bedford Park, Lawrence Manor East         Italian Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Skating Rink        Breakfast Spot                Lounge   
1                   Gym          Skating Rink           Coffee Shop   
2           Coffee Shop   Fried Chicken Joint                 Diner   
3   Japanese Restaurant                  Bank    Chinese Restaurant   
4           Coffee Shop        Sandwich Place        Cosmetics Shop   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0           Yoga Studio               Dog Run    Dim Sum Restaurant   
1                   Pub        Sandwich Place              Pharmacy   
2           Supermarket        Ice Cream Shop      Sushi Restaurant   
3           Yoga Studio          Dessert Shop                 Diner   
4             Juice Bar               Butcher                  Café   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                 Diner        Discount Store    Distribution Center  
1        Airport Lounge       Airport Service            Escape Room  
2     Mobile Phone Shop         Shopping Mall             Restaurant  
3        Discount Store   Distribution Center                Dog Run  
4      Sushi Restaurant                   Pub           Liquor Store

In [213]:
neighborhoods_venues_sorted.shape

(96, 11)

## 3.4. Cluster Neighborhoods

In [214]:
# set number of clusters
kclusters = 4

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_ 

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 2, 1, 1, 0, 1, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 3,
       1, 3, 1, 1, 1, 1, 0, 1])

In [215]:
kmeans.labels_.shape

(96,)

In [216]:
# add clustering labels
try:
    neighborhoods_venues_sorted = neighborhoods_venues_sorted.drop("Cluster Labels", 1)
except:
    print("Cluster Labels not present")
    
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
neighborhoods_venues_sorted

Cluster Labels not present


rant                       Diner   
59                      Bookstore                     Dog Run   
60                      Gastropub                 Flea Market   
61              Food & Drink Shop                 Yoga Studio   
62                    Yoga Studio              Discount Store   
63                           Café                         Pub   
64                     Restaurant                       Hotel   
65                          Trail                 Yoga Studio   
66                         Garden                Home Service   
67                    Yoga Studio                  Donut Shop   
68                            Pub                 Pizza Place   
69                 Breakfast Spot           Convenience Store   
70                     Playground                Dessert Shop   
71                     Beer Store         Fried Chicken Joint   
72                 Cosmetics Shop                Cocktail Bar   
73                     Restaurant                      Bakery   
74                           Bank           Indian Restaurant   
75                     Restaurant                        Café   
76                      Gastropub         American Restaurant   
77            American Restaurant            Sushi Restaurant   
78                    Coffee Shop                Burger Joint   
79              Health Food Store                       Trail   
80             Italian Restaurant              Ice Cream Shop   
81                    Yoga Studio                     Dog Run   
82                    Supermarket                        Bank   
83                           Café                 Salad Place   
84             Italian Restaurant         Japanese Restaurant   
85                    Coffee Shop       Portuguese Restaurant   
86                   Intersection                 Coffee Shop   
87               Doner Restaurant                Dessert Shop   
88                  Shopping Mall              Sandwich Place   
89               Doner Restaurant                Dessert Shop   
90                    Pizza Place               Shopping Mall   
91                    Pizza Place                 Coffee Shop   
92          Korean BBQ Restaurant            Doner Restaurant   
93                   Dance Studio                 Curling Ice   
94                    Yoga Studio          Dim Sum Restaurant   
95           Ethiopian Restaurant                 Escape Room   

        5th Most Common Venue        6th Most Common Venue  \
0                 Yoga Studio                      Dog Run   
1                         Pub               Sandwich Place   
2                 Supermarket               Ice Cream Shop   
3                 Yoga Studio                 Dessert Shop   
4                   Juice Bar                      Butcher   
5                    Beer Bar                     Pharmacy   
6       General Entertainment                  Yoga Studio   
7                      Bakery                      Stadium   
8                  Restaurant                Burrito Place   
9         Rental Car Location                        Plane   
10                      Diner               Discount Store   
11   Mediterranean Restaurant                Burrito Place   
12         Athletics & Sports                         Bank   
13                Salad Place              Thai Restaurant   
14                Coffee Shop           Italian Restaurant   
15                 Restaurant                         Café   
16         Italian Restaurant              Thai Restaurant   
17         Dim Sum Restaurant                        Diner   
18                        Gym          American Restaurant   
19                Coffee Shop           Italian Restaurant   
20             Sandwich Place             Department Store   
21         Turkish Restaurant             Doner Restaurant   
22        Japanese Restaurant                  Art Gallery   
23                  Pet Store                    Gift Shop   
24      

In [217]:
toronto_merged = df_toronto

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
toronto_merged = neighborhoods_venues_sorted.join(toronto_merged.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Cluster Labels                                     Neighborhood  \
0               1                                        Agincourt   
1               1                           Alderwood, Long Branch   
2               1  Bathurst Manor, Wilson Heights, Downsview North   
3               1                                  Bayview Village   
4               1                Bedford Park, Lawrence Manor East   

       1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0  Latin American Restaurant          Skating Rink        Breakfast Spot   
1                Pizza Place                   Gym          Skating Rink   
2                       Bank           Coffee Shop   Fried Chicken Joint   
3                       Café   Japanese Restaurant                  Bank   
4         Italian Restaurant           Coffee Shop        Sandwich Place   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0                Lounge           Yoga Studio               Dog Run   
1           Coffee Shop                   Pub        Sandwich Place   
2                 Diner           Supermarket        Ice Cream Shop   
3    Chinese Restaurant           Yoga Studio          Dessert Shop   
4        Cosmetics Shop             Juice Bar               Butcher   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0    Dim Sum Restaurant                 Diner        Discount Store   
1              Pharmacy        Airport Lounge       Airport Service   
2      Sushi Restaurant     Mobile Phone Shop         Shopping Mall   
3                 Diner        Discount Store   Distribution Center   
4                  Café      Sushi Restaurant                   Pub   

  10th Most Common Venue PostalCode      Borough   Latitude  Longitude  
0    Distribution Center        M1S  Scarborough  43.794200 -79.262029  
1            Escape Room        M8W    Etobicoke  43.602414 -79.543484  
2             Restaurant        M3H   North York  43.754328 -79.442259  
3                Dog Run        M2K   North York  43.786947 -79.385975  
4           Liquor Store        M5M   North York  43.733283 -79.419750

#### let's visualize the resulting clusters

In [218]:

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    print(label)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## 3.5. Examine Clusters

In [219]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(2, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
10                                Caledonia-Fairbanks                  Park   
26      East Toronto, Broadview North (Old East York)     Convenience Store   
50  Milliken, Agincourt North, Steeles East, L'Amo...                  Park   
52                        Moore Park, Summerhill East                   Gym   
61                                          Parkwoods                  Park   
65                                           Rosedale                  Park   
81      The Kingsway, Montgomery Road, Old Mill North                  Park   
94                                    York Mills West     Convenience Store   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
10         Women's Store                   Bar            Donut Shop   
26                  Park           Yoga Studio    Dim Sum Restaurant   
50            Playground          Intersection           Yoga Studio   
52                  Park      Doner Restaurant          Dessert Shop   
61                  Pool     Food & Drink Shop           Yoga Studio   
65            Playground                 Trail           Yoga Studio   
81                 River           Yoga Studio               Dog Run   
94                  Park           Yoga Studio    Dim Sum Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
10                 Diner        Discount Store   Distribution Center   
26                 Diner        Discount Store   Distribution Center   
50               Dog Run          Dessert Shop    Dim Sum Restaurant   
52    Dim Sum Restaurant                 Diner        Discount Store   
61               Dog Run          Dessert Shop    Dim Sum Restaurant   
65   Distribution Center      Department Store          Dessert Shop   
81          Dessert Shop    Dim Sum Restaurant                 Diner   
94                 Diner        Discount Store   Distribution Center   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  \
10               Dog Run      Doner Restaurant              Drugstore   
26               Dog Run      Doner Restaurant             Donut Shop   
50                 Diner        Discount Store    Distribution Center   
52   Distribution Center               Dog Run             Donut Shop   
61                 Diner        Discount Store    Distribution Center   
65    Dim Sum Restaurant                 Diner         Discount Store   
81        Discount Store   Distribution Center       Doner Restaurant   
94               Dog Run      Doner Restaurant             Donut Shop   

   PostalCode           Borough   Latitude  Longitude  
10        M6E              York  43.689026 -79.453512  
26        M4J         East York  43.685347 -79.338106  
50        M1V       Scarborough  43.815252 -79.284577  
52        M4T   Central Toronto  43.689574 -79.383160  
61        M3A        North York  43.753259 -79.329656  
65        M4W  Downtown Toronto  43.679563 -79.377529  
81        M8X         Etobicoke  43.653654 -79.506944  
94        M2P        North York  43.752758 -79.400049

In [220]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(2, toronto_merged.shape[1]))]]

Coffee Shop           Italian Restaurant   
20             Sandwich Place             Department Store   
21         Turkish Restaurant             Doner Restaurant   
22        Japanese Restaurant                  Art Gallery   
22        Japanese Restaurant                  Art Gallery   
23                  Pet Store                    Gift Shop   
24                       Bank               Baseball Field   
24                       Bank               Baseball Field   
24                       Bank               Baseball Field   
24                       Bank               Baseball Field   
25                       Bank    Middle Eastern Restaurant   
27               Liquor Store                         Café   
28              Women's Store                    Juice Bar   
29        Japanese Restaurant                   Restaurant   
30                Yoga Studio          Distribution Center   
31            Bubble Tea Shop          Japanese Restaurant   
32                Yoga Studio             Department Store   
33               Intersection                     Bus Line   
34             Medical Center                 Intersection   
35         Italian Restaurant          Fried Chicken Joint   
36  Cajun / Creole Restaurant                  Flea Market   
37                       Pool                 Dessert Shop   
38               Dessert Shop           Dim Sum Restaurant   
40                Yoga Studio             Department Store   
41                        Pub                      Brewery   
42                    Dog Run                 Dessert Shop   
43      Vietnamese Restaurant                 Dessert Shop   
44                Yoga Studio          Distribution Center   
45                   Boutique           Miscellaneous Shop   
46                Yoga Studio                      Dog Run   
47                       Bank                 Burger Joint   
48                       Café             Asian Restaurant   
49         Dim Sum Restaurant                        Diner   
51                     Bakery               Discount Store   
53               Liquor Store          Fried Chicken Joint   
54                Yoga Studio           Dim Sum Restaurant   
55       Fast Food Restaurant                        Diner   
56           Department Store                 Dessert Shop   
57       Caribbean Restaurant                          Bar   
59         Italian Restaurant                   Restaurant   
60                       Bank                 Intersection   
62             Sandwich Place        Portuguese Restaurant   
63                       Park               Breakfast Spot   
64                        Gym                Deli / Bodega   
66             Discount Store             Department Store   
67                      Diner               Discount Store   
68                Coffee Shop           Italian Restaurant   
69                 Donut Shop               Discount Store   
70         Dim Sum Restaurant                        Diner   
71       Fast Food Restaurant               Sandwich Place   
72             Clothing Store          American Restaurant   
73         Italian Restaurant                          Pub   
74          Electronics Store                  Pizza Place   
75        Japanese Restaurant           Italian Restaurant   
76                     Bakery                      Brewery   
77                        Pub                         Bank   
78          Indian Restaurant                  Flower Shop   
79        Distribution Center             Department Store   
80     Furniture / Home Store                    Bookstore   
82               Burger Joint                  Coffee Shop   
83                 Restaurant           Seafood Restaurant   
84                        Bar                    Bookstore   
85               Hockey Arena          Distribution Center   
86             Sandwich Place               Discount Store   
88                Auto Garage                  Yog

In [221]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(2, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
39                                   Humberlea, Emery        Baseball Field   
58  Old Mill South, King's Mill Park, Sunnylea, Hu...        Baseball Field   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
39            Donut Shop    Dim Sum Restaurant                 Diner   
58            Donut Shop    Dim Sum Restaurant                 Diner   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
39        Discount Store   Distribution Center               Dog Run   
58        Discount Store   Distribution Center               Dog Run   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  \
39      Doner Restaurant           Yoga Studio                   Farm   
58      Doner Restaurant           Yoga Studio                   Farm   

   PostalCode     Borough   Latitude  Longitude  
39        M9M  North York  43.724766 -79.532242  
58        M8Y   Etobicoke  43.636258 -79.498509

In [222]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(2, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
87                   Weston                  Park           Yoga Studio   
89  Willowdale, Newtonbrook                  Park           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
87      Doner Restaurant          Dessert Shop    Dim Sum Restaurant   
89      Doner Restaurant          Dessert Shop    Dim Sum Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
87                 Diner        Discount Store   Distribution Center   
89                 Diner        Discount Store   Distribution Center   

   9th Most Common Venue 10th Most Common Venue PostalCode     Borough  \
87               Dog Run             Donut Shop        M9N        York   
89               Dog Run             Donut Shop        M2M  North York   

     Latitude  Longitude  
87  43.706876 -79.518188  
89  43.789053 -79.408493